add aria label for prop to remove button

britt6612 opened the pull request:
<!--- Provide a general summary of the PR in the Title above -->

#### What does this PR do?
adds a prop for `aria-label` to be passed in
#### Where should the reviewer start?
tag.js
#### What testing has been done on this PR?
locally
#### How should this be manually tested?
locally
#### Do Jest tests follow these best practices?
I didnt add one since it got added to other snapshots
- [ ] `screen` is used for querying.
- [ ] The correct query is used. (Refer to [this list of queries](https://testing-library.com/docs/queries/about/#priority))
- [ ] `asFragment()` is used for snapshot testing.

#### Any background context you want to provide?
On the Tag component there is no way for the implementer to provide an aria-label
#### What are the relevant issues?
closes #7555 
#### Screenshots (if appropriate)

#### Do the grommet docs need to be updated?
yes
#### Should this PR be mentioned in the release notes?
yes
#### Is this change backwards compatible or is it a breaking change?
backwards compatible

taysea commented on src/js/components/Tag/Tag.js:
```diff
@@ -9,7 +9,10 @@ import { StyledRemoveButton, StyledTagButton } from './StyledTag';
 const Tag = forwardRef(
-  ({ name, value, size = 'medium', onRemove, onClick, ...rest }, ref) => {
+  (
+    { name, value, size = 'medium', onRemove, onClick, ariaLabel, ...rest },
```
This approach seems like it would apply to the tag, not the subcomponent. Wondering if we should support

```
removeProps= { 'aria-label': '' }
```

For now I'd say we only let aria-label be the prop in there.

MikeKingdom replied:
I would agree with this and in fact given how `rest` is put on the container you'd potentially break existing usage of ariaLabel by apps.

taysea commented on src/js/components/Tag/Tag.js:
```diff
@@ -66,6 +69,7 @@ const Tag = forwardRef(
         {onRemove && (
           <StyledRemoveButton
             onClick={onRemove}
+            aria-label={ariaLabel || 'Remove tag'}
```
If we add "Remove tag" as a string, we should include it in default.json message and allow for internationalization.

taysea replied:
I also wonder if instead of "Remove tag" as the fallback if it should be `Remove ${name} ${value}`

britt6612 replied:
good thinking yeah I can do that

MikeKingdom replied:
Note that to properly localize this it would be no `$` in default.json and the format call would be passed the values:
In default.json
```json
"tag": {
  "removeLabel": "Remove {name} {value}"
},
```
And in Tag.js something like:
```jsx
format({
  id: 'tag.removeLabel',
  messages,
  values: {
    name,
    value
  },
})
```
And you have to document it for Grommet's index.d.ts/prop-types.

Typically we've also added a messages prop to the component so that it can be locally set/overridden (not just globally set in Grommet.) Not sure that's always required.

MikeKingdom commented on src/js/components/Tag/Tag.js:
```diff
@@ -47,7 +67,7 @@ const Tag = forwardRef(
     }
 
-    const removeProps = !theme.tag.remove.kind
+    const computedRemoveProps = !theme.tag.remove.kind
```
seems like  better name for this would have "theme" in it.

britt6612 replied:
yeah I was wondering on the naming thank you!

britt6612 replied:
I changed to `themeTagRemoveProps`

taysea commented on src/js/components/Tag/Tag.js:
```diff
@@ -7,12 +7,32 @@ import { Text } from '../Text';
+      onRemove,
+      onClick,
+      messages,
+      removeProps = { 'aria-label': '' },
```
```suggestion
```
Do we even need this anymore since it's handled my `messages`?

britt6612 replied:
aw yes good catch

taysea commented on src/js/components/Tag/Tag.js:
```diff
@@ -47,7 +66,7 @@ const Tag = forwardRef(
     }
 
-    const removeProps = !theme.tag.remove.kind
+    const themeTagRemoveProps = !theme.tag.remove.kind
```
I don't think we need to rename this anymore

britt6612 replied:
aw yes will revert

taysea commented on src/js/components/Tag/Tag.js:
```diff
@@ -66,7 +85,9 @@ const Tag = forwardRef(
           <StyledRemoveButton
             onClick={onRemove}
+            {...themeTagRemoveProps}
             {...removeProps}
```
Related to above comment, but `removeProps` was renamed and now no longer exists. Getting an error locally for missing variable.

I'd say we don't even need to renamed `removeProps` anymore, but can delete "themeTagRemoveProps"

taysea commented on src/js/components/Tag/__tests__/__snapshots__/Tag-test.tsx.snap:
```diff
@@ -551,6 +551,7 @@ exports[`Tag onRemove 1`] = `
       </span>
     </div>
     <button
+      aria-label="Remove undefined Test"
```
If name doesn't exist, I wouldn't expect it to include "undefined" in the aria-label. I'd expect this to read "Remove Test"

taysea commented on src/js/components/Data/__tests__/__snapshots__/Data-test.tsx.snap:
```diff
@@ -400,7 +400,7 @@ exports[`Data controlled search 1`] = `
       class="c11"
     >
-      4 results of undefined items
+      4 results of  items
```
This read a little odd. It feels like ideally if the total wasn't provided it'd read either:
- "4 results" OR
- "4 results of unknown total items" (or something along those lines)

@MikeKingdom you're more expert in localization -- is this something we should worry about or accept this change?

MikeKingdom replied:
Yeah. In general we should never put "undefined" or "null" in a message. But we should also make sure we're not creating a message that isn't "friendly". Typically in this situation the guidance is if you don't know the total then only say "4 results".

MikeKingdom commented on src/js/languages/default.json:
```diff
@@ -161,6 +161,9 @@
     "tabContents": "Tab Contents"
   },
+  "tag": {
+    "removeLabel": "Remove {name} {value}"
```
This probably is a bit simplistic. I think we probably need 3 variations of this message based on whether name or value is defined or both. And the ':' is there in the visible rendering for a reason. Without it these can get confusing.

MikeKingdom commented on src/js/components/Tag/__tests__/__snapshots__/Tag-test.tsx.snap:
```diff
@@ -551,6 +551,7 @@ exports[`Tag onRemove 1`] = `
       </span>
     </div>
     <button
+      aria-label="Remove Test"
```
We should probably have a test case for the other 2 variations

britt6612 replied:
will do

MikeKingdom commented on src/js/components/Tag/index.d.ts:
```diff
@@ -5,6 +5,9 @@ export interface TagProps {
   onClick?: (...args: any[]) => any;
   onRemove?: (...args: any[]) => any;
+  messages?: {
+    removeLabel?: string;
```
this needs to be updated for the new keys

MikeKingdom commented on src/js/components/Tag/propTypes.js:
```diff
@@ -7,6 +7,9 @@ if (process.env.NODE_ENV !== 'production') {
     onClick: PropTypes.func,
     onRemove: PropTypes.func,
+    messages: PropTypes.shape({
+      removeLabel: PropTypes.string,
```
same

MikeKingdom reviewed (commented):
Thanks Brittany. Looks like it needs some types updates.

britt6612 commented:
@MikeKingdom 

glad I added the tests. I saw that `value` is actually required so I changed the json file to just be 2 use cases. 
```
  "tag": {
    "removeLabel": {
      "nameAndValue": "Remove {name}: {value}",
      "valueOnly": "Remove {value}"
    }
  },
```
since `value` is always required we only need `valueOnly` and `nameAndValue` So updates propTypes, TS and tests.

MikeKingdom commented:
> since `value` is always required we only need `valueOnly` and `nameAndValue` So updates propTypes, TS and tests.

Fair enough although I think that's a silly requirement and people tend to pick one or the other based on whether or not they want it to be bold.

jcfilben reviewed (approved):
Looks good!

jcfilben commented:
Holding off on merging until after the upcoming patch release

jcfilben commented:
Updated with master but looks like we need to bump the bundlesize up
